deltaconcor/src/main/java/uk/gov/justice/laadces/deltaconcor/Runner.java

```diff
@@ -0,0 +1,353 @@
+            nextLowId = sliceHighId + 1;
+            log.info("visitForwards: {} rows with {} <= concorId <= {}", contentLen, sliceLowId, sliceHighId);
+            for (final ConcorContribution concorContribution : content) {
+                if (!visitor.visit(concorContribution)) {
+                    return concorContribution.id();
+                }
+            }
+        } while (slice.hasNext() && contentLen > 0 && nextLowId <= endId);
+        return endId;
+    }
+
+    /**
+     * Uses a slower query to visit backwards through the ConcorContribution table; however this method works even if
+     * there are more than 1000 maatIds to back-fill.
+     *
+     * @param startId start ID to walk back from
+     * @param endId end ID to finish walking back at
+     * @param visitor the RecordVisitor to call for each record visited
+     * @return the ID of the last record visited
+     */
+    private long slowVisitBackwards(final long startId, final long endId, final RecordVisitor visitor) {
+        assert startId >= endId;
+
+        long nextHighId = startId;
+        Slice<ConcorContribution> slice;
+        List<ConcorContribution> content;
+        int contentLen;
+        long sliceLowId, sliceHighId;
+
+        do {
+            slice = concorRepository.findByIdBetweenAndStatusOrderByIdDesc(
+                    endId, nextHighId, "SENT", Pageable.ofSize(BATCH_SIZE));
+            content = slice.getContent();
+            contentLen = content.size();
+            sliceHighId = contentLen > 0 ? content.get(0).id() : nextHighId;
+            sliceLowId = contentLen > 0 ? content.get(contentLen - 1).id() : endId;
+            nextHighId = sliceLowId - 1L;
+            log.info("slowVisitBackwards: {} rows with {} >= concorId >= {}", contentLen, sliceHighId, sliceLowId);
+            for (final ConcorContribution concorContribution : content) {
+                if (!visitor.visit(concorContribution)) {
+                    return concorContribution.id();
+                }
+            }
+        } while (slice.hasNext() && contentLen > 0 && nextHighId >= endId);
+        return endId;
+    }
+
+    /**
+     * Uses a faster query to visit backwards through the ConcorContribution table; however it only works if there are
+     * less than 1000 maatIds to back-fill.
+     *
+     * @param startId start ID to walk back from
+     * @param endId end ID to finish walking back at
+     * @param visitor the RecordVisitor to call for each record visited
+     * @return the ID of the last record visited
+     */
+    private long fastVisitBackwards(final long startId, final long endId, final RecordVisitor visitor) {
+        assert startId >= endId;
+
+        long nextHighId = startId;
+        Slice<ConcorContribution> slice;
+        List<ConcorContribution> content;
+        int contentLen;
+        long sliceLowId, sliceHighId;
+
+        do {
+            slice = concorRepository.findByIdBetweenAndStatusAndRepIdInOrderByIdDesc(
```

Comment:
Not asking to change, but asking out of curiosity, could you have used the fast query in every case, by sending chunks of 999 MAAT IDs at a time?

Reply:
Maybe, but then I'd have to split the maatIds into chunks, then merge the results again, so I think it would have made the code considerably more complicated. It might be even quicker to do that, but 5 minutes per month was good enough.